Source/GSAutoLayoutVFLParser.m

```diff
@@ -0,0 +1,756 @@
+/* Copyright (C) 2022 Free Software Foundation, Inc.
+   
+   By: Benjamin Johnson
+   Date: 11-11-2022
+   This file is part of the GNUstep Library.
+   
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+   
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+   
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free
+   Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+   Boston, MA 02110 USA.
+*/
+
+#import "GSAutoLayoutVFLParser.h"
+#import <AppKit/AppKit.h>
+
+struct GSObjectOfPredicate
+{
+  NSNumber *priority;
+  NSView *view;
+  NSLayoutRelation relation;
+  CGFloat constant;
+};
+typedef struct GSObjectOfPredicate GSObjectOfPredicate;
+
+NSInteger const GS_DEFAULT_VIEW_SPACING = 8;
+NSInteger const GS_DEFAULT_SUPERVIEW_SPACING = 20;
+
+@implementation GSAutoLayoutVFLParser
+
+- (instancetype) initWithFormat:(NSString *)format
+                       options:(NSLayoutFormatOptions)options
+                       metrics:(NSDictionary *)metrics
+                         views:(NSDictionary *)views
+{
+  if (self = [super init])
+    {
+      if ([format length] == 0)
+        {
+          [self failParseWithMessage:@"Cannot parse an empty string"];
```

Comment:
In GNUstep we normally have a space after the colon.

I am not sure whether I like to see an exception being thrown from the init method. IN most cases this should result in the object not being cleaned up. I am not sure if even ARC would handle that correctly.

Reply:
Regarding the potential memory leak, I did some research on this and your concern is correct however this is to be excepted when using NSException. From Apple's docs "The Cocoa frameworks are generally not exception-safe. The general pattern is that exceptions are reserved for programmer error only, and the program catching such an exception should quit soon afterwards." https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Exceptions/Articles/ExceptionsAndCocoaFrameworks.html#//apple_ref/doc/uid/TP40009045-SW1

Reply:
Here I disagree with you. At least in the old times you would have returned `nil` here and as usual the caller has to handle that. Take a look at the documentation cited in the second answer here: https://stackoverflow.com/questions/6574938/why-not-throw-an-exception-if-super-init-returns-nil